### PR TITLE
MP-9464 - Generate Default Integration Events

### LINKED ISSUE #821: [Nox.Generator] Generate Default Integration Events
- relates to #799 
- depends on #820 

1. reuse the Custom IE template, get the attributes from the relate Domain Event
1. generate IE only when  Entity Persistence RaiseEvents is DomainAndIntegrationEvents
1. Create a template to generate a Notification Handler to consume the DomainEvent and add the IE  to the MessagesOutbox
1. Add the template to Nox.Generator.Application.NotificationHandlers
1. generated  file name \<Service\>.Application.NotificationHandlers.\<DomainEventHandler\>.g.cs
2. create an example on Client API (Yaml), so a IE is generated (when done send this information to @Mornevanzyl  to be added to the documentation Extending Nox)

### PR DESCRIPTION
**Pull request by dejobratic**
- Added DefaultIntegrationEvent template and generator, that generates IEs only for entities that have integration events;
  - Since IEs have DTO properties, added extension methods for mapping compound Nox types to their DTOs and for mapping entities to DTOs;
- Renamed existing IntegrationEvent template and generator to CustomIntegrationEvent template and generator;
- Added DomainEventHandlers template and generator that generates handlers only for entities that have domain events;

Apologies for the huge number of files. However, 10ish percent of them are the ones I've actually worked on - the others are generated files.

### CODE REVIEW

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        _outboxRepository = outboxRepository;
+    }
+
+    public virtual async Task Handle({{entity.Name}}{{crudOperation}} domainEvent, CancellationToken cancellationToken)
+    {
+{{- if raiseIntegrationEvent }}
+        var dto = domainEvent.{{entity.Name}}.ToDto();
+        var @event = new IntegrationEvents.{{entity.Name}}{{crudOperation}}(dto);
+        await RaiseIntegrationEventAsync(@event);
```
Comment (rochar) [inline src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs]: I would keep complexity low on this, seems unnecessary 
```suggestion
     await _outboxRepository.AddAsync(@event);
```

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    protected {{className}}Base(IOutboxRepository outboxRepository)
+    {
+        _outboxRepository = outboxRepository;
+    }
+
+    public virtual async Task Handle({{entity.Name}}{{crudOperation}} domainEvent, CancellationToken cancellationToken)
+    {
+{{- if raiseIntegrationEvent }}
+        var dto = domainEvent.{{entity.Name}}.ToDto();
+        var @event = new IntegrationEvents.{{entity.Name}}{{crudOperation}}(dto);
```
Comment (rochar) [inline src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs]: for avoid conflicts add the full namespace
```suggestion
        var @event = new {{codeGeneratorState.??}}.IntegrationEvents.{{entity.Name}}{{crudOperation}}(dto);
```

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    private readonly IOutboxRepository _outboxRepository;
+
+    protected {{className}}Base(IOutboxRepository outboxRepository)
+    {
+        _outboxRepository = outboxRepository;
+    }
+
+    public virtual async Task Handle({{entity.Name}}{{crudOperation}} domainEvent, CancellationToken cancellationToken)
+    {
+{{- if raiseIntegrationEvent }}
```
Comment (rochar) [inline src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs]: the should be on the generator, and we would not generate a file / handler that is not needed

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+        foreach (var (crudOperation, raiseIntegrationEvent, entity) in GroupEntitiesWithDomainEventsByCrudOperation(codeGeneratorState.Solution.Domain.Entities))
+        {
+            context.CancellationToken.ThrowIfCancellationRequested();
+
+            new TemplateCodeBuilder(context, codeGeneratorState)
+                .WithClassName($"{entity.Name}{crudOperation}DomainEventHandler")
+                .WithFileNamePrefix($"Application.DomainEventHandlers")
+                .WithObject("crudOperation", crudOperation)
+                .WithObject("raiseIntegrationEvent", raiseIntegrationEvent)
```
Comment (rochar) [inline src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs]: skip if we are not raising this object is unnecessary

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                .WithObject("entity", entity)
+                .GenerateSourceCodeFromResource("Application.DomainEventHandlers.DomainEventHandler");
+        }
+    }
+
+    private IEnumerable<(string CrudOperation, bool RaiseIntegrationEvent, Entity entity)> GroupEntitiesWithDomainEventsByCrudOperation(IEnumerable<Entity> entities)
+    {
+        var entitiesWithDomainEvents = GetEntitiesThatHaveDomainEvents(entities);
+
+        return entitiesWithDomainEvents.Where(e => e.Persistence?.Create?.RaiseDomainEvents == true).Select(e => ("Created", e.Persistence?.Create?.RaiseIntegrationEvents ?? false, e))
```
Comment (rochar) [inline src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs:40]: we could basically exclude the ones that are not raising IE?

File: src/Nox.Generator/Application/Extensions/EntityExtensions.template.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        SetIfNotNull(entity?.{{relationship.Name}}, () => dto.{{relationship.Name}} = entity!.{{relationship.Name}}.Select(e => e.ToDto()).ToList());
+    {{- else}}
+        SetIfNotNull(entity?.{{relationship.Name}}, () => dto.{{relationship.Name}} = entity!.{{relationship.Name}}!.ToDto());
+    {{-end}}
+{{- end }}
+
+        return dto;
+    }
+
+    private static void SetIfNotNull(object? value, Action setPropertyAction)
```
Comment (rochar) [inline src/Nox.Generator/Application/Extensions/EntityExtensions.template.cs]: we have this on the FluentApiUtil.... but we do not have a good project to have this to be reused by everyone,... I do not have a better Idea for now, I would juts add a FluentApiExtensions to Noc.Lib and reuse it

**Review by rochar — CHANGES_REQUESTED** [review]

File: src/Nox.Generator/Application/IntegrationEvents/DefaultIntegrationEvent.template.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+using Nox.Application;
+
+using {{codeGeneratorState.ApplicationNameSpace}}.Dto;
+
+namespace {{codeGeneratorState.ApplicationNameSpace}}.IntegrationEvents;
+
+/// <summary>
+/// {{entity.Name}}{{crudOperation}} integration event.
+/// </summary>
+internal record {{entity.Name}}{{crudOperation}}({{entity.Name}}Dto {{entity.Name}}) :  IIntegrationEvent;
```
Comment (rochar) [inline src/Nox.Generator/Application/IntegrationEvents/DefaultIntegrationEvent.template.cs:15]: just a suggestion are y ok in renaming in this PR crudOperation to operation?

**Review by rochar — CHANGES_REQUESTED** [review]

File: src/Nox.Generator/Application/IntegrationEvents/DefaultIntegrationEvent.template.cs (hunk 7 above)
Comment (dejobratic) [reply src/Nox.Generator/Application/IntegrationEvents/DefaultIntegrationEvent.template.cs:15]: Changes were applied.

File: src/Nox.Generator/Application/Extensions/EntityExtensions.template.cs (hunk 6 above)
Comment (dejobratic) [reply src/Nox.Generator/Application/Extensions/EntityExtensions.template.cs]: Changes were applied. Thanks for the suggestion.

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs (hunk 5 above)
Comment (dejobratic) [reply src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs:40]: Changes were applied.

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs (hunk 4 above)
Comment (dejobratic) [reply src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs]: Changes were applied.

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs (hunk 3 above)
Comment (dejobratic) [reply src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs]: Changes were applied.

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs (hunk 2 above)
Comment (dejobratic) [reply src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs]: Changes were applied.

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs (hunk 1 above)
Comment (dejobratic) [reply src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandler.template.cs]: Changes were applied.

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+                    .WithClassName($"{entity.Name}{operation}DomainEventHandler")
+                    .WithFileNamePrefix($"Application.DomainEventHandlers")
+                    .WithObject("operation", operation)
+                    .WithObject("entity", entity)
+                    .GenerateSourceCodeFromResource("Application.DomainEventHandlers.DomainEventHandler");
+            }
+        }
+    }
+
+    private IEnumerable<(string Operation, bool RaiseIntegrationEvent, Entity entity)> GroupEntitiesWithDomainEventsByCrudOperation(IEnumerable<Entity> entities)
```
Comment (rochar) [inline src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs]:
```suggestion
    private IEnumerable<(string Operation, bool RaiseIntegrationEvent, Entity entity)> GroupEntitiesWithDomainEventsByOperation(IEnumerable<Entity> entities)
```

**Review by rochar — DISMISSED** [review]
add a simple suggestion

File: src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs (hunk 8 above)
Comment (dejobratic) [reply src/Nox.Generator/Application/DomainEventHandlers/DomainEventHandlerGenerator.cs]: Missed that one. It's updated now. Thanks.

**Review by rochar — APPROVED** [review]

**Review by gjorgji-grgovski-authoritypartners — APPROVED** [review]

**Pull request merged**